open-craft/openedx-completion-aggregator
# Add service to periodically aggregate out of date completions.

**Description:** 

This provides a service that runs out-of-band and updates out-of-date completion aggregators.
When a completion would previously have called the aggregator task directly, it now marks its BlockCompletions out-of-date in a StaleCompletion table.  When the service runs, it gets all the stale blocks from the table, runs one completion task for each enrollment that is out of date, and marks all its stale completions resolved.  Later, resolved completions are deleted from the table, but we don't do this immediately to reduce table churn, and to provide a little visibility into recent table activity.


**JIRA:** OC-3968

**Dependencies:** None

**Merge deadline:** None

**Installation instructions:** None.

**Testing instructions:**
1. Run the management command.  ./manage.py run_aggregator_service -v2
2. Create some BlockCompletions, either by browsing a course, or from a ./manage.py shell
3. Verify (from the logging or from the command line) that StaleCompletions are being created and resolved.  
4. If running from edx-platform, verify that running a celery task worker causes the aggregators to actually be updated.  If running from the openedx-completion-aggregator repo, verify that running a celery task worker causes tasks to be taken off the queue, but fail, because the modulestore can't be accessed.

**Reviewers:**

**Merge checklist:**

**Post merge:**

**Author concerns:** We still need to make sure this service gets run from a cronjob when deployed to 
McKinsey.  I'm not sure how we do that.
